ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -369,8 +372,8 @@ let revert_vm_fields ~__context ~snapshot ~vm =
     if post_MNR
     then do_not_copy
     else extended_do_not_copy in
+  let snap_metadata = ensure_domain_type_is_consistent ~__context ~snap_metadata in
```

Comment:
I'm not sure that moving it improves things. Now you need to refer to fields by string name, that is not typechecked.
Previously you could refer to HVM_boot_policy by using get/set_HVM_boot_policy.

Reply:
There are three cases, which Richard's patch implements:
1. The snapshot metadata includes `domain_type` which is not `unspecified`. All is good.
2. The snapshot metadata includes `domain_type`, but it has the value `unspecified`. We need to use `Xapi_vm_helpers.derive_domain_type` to set the `domain_type`.
3. The snapshot metadata does not include `domain_type`. We need to use `Xapi_vm_helpers.derive_domain_type` to set the `domain_type`.

This move was to detect case 3. The old `ensure_domain_type_is_consistent` function didn't have this information, because `copy_vm_fields` only overrides fields that are present in the snapshot metadata, and there is scope for stale metadata to be left behind. However, having to use the untyped data is not great.

Another option is to leave the `ensure_domain_type_is_consistent` call where it was, but add `snap_metadata` as an argument to it. Then it can still do the checks to detect case 2 and 3, while using the normal DB API to do the update. I don't think that we can't get away from checking for the untyped string `"unspecified"` in the snapshot metadata though.